File: frontend/ts/types/organisation.ts

import type { DateString, DateTimeString, SecteurActiviteOut, UUID } from "./base";

export type StatutOrganisation = "en_attente" | "active" | "inactive";

export type TypeOrganisation =
  | "entreprise"
  | "ong"
  | "institution_publique"
  | "startup"
  | "association";

export type RoleOrganisation = "employe" | "administrateur_page";


export interface OrganisationOut {
  id: UUID;
  nom_organisation: string;
  slug: string;
  type_organisation: TypeOrganisation;
  secteur_activite: SecteurActiviteOut | null;
  adresse: string | null;
  ville: string | null;
  pays: string | null;
  pays_nom: string | null;
  site_web: string | null;
  email_general: string | null;
  telephone_general: string | null;
  logo: string | null;
  description: string | null;
  date_creation: DateString | null;
  statut: StatutOrganisation;
  est_suivi: boolean;
  nombre_abonnes: number;
  nombre_membres: number;
}


export interface MembreOrganisationOut {
  id: UUID;
  profil: UUID;
  organisation: UUID;
  role_organisation: RoleOrganisation;
  poste: UUID | null;
  est_actif: boolean;
  date_joindre: DateString;
}


export interface AbonnementOrganisationOut {
  profil: UUID;
  organisation: UUID;
  date_abonnement: DateTimeString;
}

export interface OrganisationCreate {
  nom_organisation: string;
  type_organisation: TypeOrganisation;
  secteur_activite: UUID | null;
  adresse: string | null;
  ville: string | null;
  pays: string | null;
  email_general: string | null;
  telephone_general: string | null;
  description: string | null;
  date_creation: DateString | null;
}

export interface OrganisationUpdate {
  nom_organisation?: string | null;
  type_organisation?: TypeOrganisation | null;
  secteur_activite?: UUID | null;
  adresse?: string | null;
  ville?: string | null;
  pays?: string | null;
  email_general?: string | null;
  telephone_general?: string | null;
  description?: string | null;
  date_creation?: DateString | null;
  statut?: StatutOrganisation | null;
}

export interface MembreAdd {
  profil_id: UUID;
  role_organisation: RoleOrganisation;
  poste?: UUID | null;
}

export interface MembreUpdate {
  role_organisation?: RoleOrganisation | null;
  poste?: UUID | null;
}

File: network/models/organisation.py
from django.db import models
from django.utils.translation import gettext_lazy as _
from core.models import ENSPMHubBaseModel
from django_countries.fields import CountryField



class Organisation(ENSPMHubBaseModel):
    """Représente les structures partenaires (Entreprises, Startups, etc.)"""
    TYPE_CHOICES = [
        ('entreprise', 'Entreprise'), ('ong', 'ONG'), ('startup', 'Startup'),
        ('institution_publique', 'Institution publique'), ('universite', 'Université'),
        ('association', 'Association'), ('autre', 'Autre'),
    ]
    STATUT_CHOICES = [('active', 'Active'), ('inactive', 'Inactive'), ('en_attente', 'En attente')]

    nom_organisation = models.CharField(max_length=255, verbose_name=_("Nom"))
    slug = models.SlugField(unique=True, db_index=True)
    type_organisation = models.CharField(max_length=30, choices=TYPE_CHOICES)
    secteur_activites = models.ManyToManyField('core.SecteurActivite')
    
    logo = models.ImageField(upload_to='network/orgs/logos/', null=True, blank=True)
    description = models.TextField(null=True, blank=True)
    site_web = models.URLField(null=True, blank=True)
    
    ville = models.CharField(max_length=100, null=True, blank=True)
    pays = CountryField(null=True, blank=True)
    adresse = models.CharField(max_length=255, null=True, blank=True)
    
    statut = models.CharField(max_length=20, choices=STATUT_CHOICES, default='en_attente')

    class Meta:
        db_table = 'network_organisation'
        verbose_name = _("Organisation")

    def __str__(self):
        return self.nom_organisation


class MembreOrganisation(ENSPMHubBaseModel):
    """Utilisateurs qui sont membres d'une organisation"""
    ACCES_CHOICES = [
        ('admin', 'Administrateur'), 
        ('membre', 'Membre')
    ]
    
    profil = models.ForeignKey('users.Profil', on_delete=models.CASCADE, related_name='membres_orgs')
    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE, related_name='membres')
    date_membre = models.DateTimeField(auto_now_add=True)
    acces = models.CharField(max_length=20, choices=ACCES_CHOICES, default='membre')

    class Meta:
        db_table = 'network_membre_organisation'
        unique_together = ('profil', 'organisation')


class AbonnementOrganisation(ENSPMHubBaseModel):
    """Utilisateurs qui suivent une page organisation (Followers)"""
    profil = models.ForeignKey('users.Profil', on_delete=models.CASCADE, related_name='abonnements_orgs')
    organisation = models.ForeignKey(Organisation, on_delete=models.CASCADE, related_name='abonnes')
    date_abonnement = models.DateTimeField(auto_now_add=True)

    class Meta:
        db_table = 'network_abonnement_organisation'
        unique_together = ('profil', 'organisation')